src/server/gc_worker/compaction_filter.rs

```diff
@@ -337,86 +405,75 @@ impl Drop for WriteCompactionFilter {
-        _: &mut bool,
-    ) -> bool {
-        self.near_seek_distance += 1;
-        let (key_prefix, commit_ts) = split_ts(key);
-        if commit_ts > self.safe_point {
-            return false;
+        value_type: CompactionFilterValueType,
+    ) -> CompactionFilterDecision {
+        if self.is_invalid {
+            // If there are already some errors, do nothing.
+            return CompactionFilterDecision::Keep;
         }
 
-        self.versions += 1;
-        if self.key_prefix != key_prefix {
-            self.key_prefix.clear();
-            self.key_prefix.extend_from_slice(key_prefix);
-            self.remove_older = false;
-            self.switch_key_metrics();
-        }
-
-        let mut filtered = self.remove_older;
-        let write = parse_write(value);
-        if !self.remove_older {
-            match write.write_type {
-                WriteType::Rollback | WriteType::Lock => filtered = true,
-                WriteType::Put => self.remove_older = true,
-                WriteType::Delete => {
-                    self.remove_older = true;
-                    filtered = true;
-                    self.handle_delete_mark(key);
-                }
+        match self.do_filter_v2(level, key, value, value_type) {
+            Ok(decision) => decision,
+            Err(e) => {
+                warn!("compaction filter meet error: {}", e);
+                self.is_invalid = true;
+                CompactionFilterDecision::Keep
             }
         }
-
-        if filtered {
-            self.handle_filtered_write(write);
-            self.filtered += 1;
-        }
-
-        filtered
     }
 }
 
-fn split_ts(key: &[u8]) -> (&[u8], u64) {
+fn split_ts(key: &[u8]) -> Result<(&[u8], u64), String> {
     match Key::split_on_ts_for(key) {
-        Ok((key, ts)) => (key, ts.into_inner()),
-        Err(_) => panic!("invalid write cf key: {}", hex::encode_upper(key)),
+        Ok((key, ts)) => Ok((key, ts.into_inner())),
+        Err(_) => Err(format!("invalid write cf key: {}", hex::encode_upper(key))),
     }
 }
 
-fn truncate_ts(key: &[u8]) -> &[u8] {
+fn truncate_ts(key: &[u8]) -> Result<&[u8], String> {
     match Key::truncate_ts_for(key) {
-        Ok(prefix) => prefix,
-        Err(_) => panic!("invalid write cf key: {}", hex::encode_upper(key)),
+        Ok(prefix) => Ok(prefix),
+        Err(_) => Err(format!("invalid write cf key: {}", hex::encode_upper(key))),
     }
 }
 
-fn parse_write(value: &[u8]) -> WriteRef {
+fn parse_write(value: &[u8]) -> Result<WriteRef, String> {
     match WriteRef::parse(value) {
-        Ok(write) => write,
-        Err(_) => panic!("invalid write cf value: {}", hex::encode_upper(value)),
+        Ok(write) => Ok(write),
+        Err(_) => Err(format!(
+            "invalid write cf value: {}",
+            hex::encode_upper(value)
+        )),
     }
 }
 
+fn too_many_delete_marks(props: &MvccProperties) -> bool {
+    props.num_deletes as f64 / props.num_versions as f64 > 0.2
```

Comment:
I think deletes has a  higher priority  than updates because a little scan   request (such as order by xx limit 1) may  read a lot of versions or tombstone, but updates will not cause this problem. So please set the default ratio  by 10%.

Reply:
Can we make this parameter configurable?